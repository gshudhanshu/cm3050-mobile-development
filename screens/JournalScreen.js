import {
  StyleSheet,
  Text,
  View,
  SafeAreaView,
  ScrollView,
  KeyboardAvoidingView,
  StatusBar,
  TextInput,
  TouchableOpacity,
} from 'react-native'
import React, { useEffect } from 'react'
import { RFValue } from 'react-native-responsive-fontsize'
import GlobalStyles from '../utils/GlobalStyles'
import theme from '../utils/theme'
import CText from '../components/common/CText'
import Header from '../components/Header'

import { useNavigation } from '@react-navigation/native'
import useJournalStore from '../store/useJournalStore'
import useAuthStore from '../store/useAuthStore'
import JournalCard from '../components/JournalCard'

import dayjs from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import timezone from 'dayjs/plugin/timezone'
dayjs.extend(customParseFormat)
dayjs.extend(timezone)
dayjs.tz.setDefault('Europe/London')

export default function JournalScreen() {
  const { journals, fetchJournals } = useJournalStore()
  const { user } = useAuthStore()
  const navigation = useNavigation()

  useEffect(() => {
    if (user?.uid) {
      fetchJournals(user.uid)
    }
  }, [user?.uid, fetchJournals])

  const addNewJournal = () => {
    navigation.navigate('JournalEntry')
  }

  return (
    <SafeAreaView style={GlobalStyles.safeAreaContainer}>
      <KeyboardAvoidingView style={GlobalStyles.container}>
        <ScrollView contentContainerStyle={styles.scrollViewContent}>
          <StatusBar
            barStyle='light-content'
            backgroundColor={theme.colors.primary}
          />
          <Header showBack={false} useLogo={false} title={'Journal'} />
          {/* Search input */}
          <View style={styles.searchInputContainer}>
            <TextInput placeholder='Search...' style={[GlobalStyles.input]} />
          </View>
          {/* Journal cards */}
          <View style={styles.journalContainer}>
            <View style={GlobalStyles.blockContainer}>
              <CText weight='semiBold' style={GlobalStyles.blockTitle}>
                Browse by Journals
              </CText>
              <TouchableOpacity
                style={[GlobalStyles.button]}
                onPress={addNewJournal}
              >
                <CText style={[GlobalStyles.buttonText]}>Add Journal</CText>
              </TouchableOpacity>
            </View>
            <View style={styles.journalsContainer}>
              {journals.map((journal) => (
                <JournalCard
                  key={journal.id}
                  imageUri={journal.imageUrl || 'https://placehold.it/300x300'}
                  date={dayjs(journal.createdAt).format('MMMM DD, YYYY')}
                  title={journal.title}
                  description={journal.description}
                  onButtonPress={() =>
                    navigation.navigate('JournalDetail', {
                      ...journal,
                    })
                  }
                />
              ))}
            </View>
          </View>
        </ScrollView>
      </KeyboardAvoidingView>
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  container: {
    justifyContent: 'flex-start',
    flex: 1,
  },
  scrollViewContent: {
    paddingBottom: RFValue(80),
  },
  searchInputContainer: {
    marginVertical: RFValue(10),
  },

  journalsContainer: {
    gap: RFValue(20),
  },
})
