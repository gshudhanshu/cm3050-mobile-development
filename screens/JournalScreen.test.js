// // JournalScreen.test.js

// import React from 'react'
// import { render } from '@testing-library/react-native'
// import JournalScreen from './JournalScreen' // Adjust the import path as needed.
// import '../store/useJournalStore'

// // Continue to mock necessary dependencies as previously shown
// jest.mock('@react-navigation/native', () => ({
//   useNavigation: () => ({
//     navigate: jest.fn(),
//   }),
// }))

// jest.mock('../store/useJournalStore', () => ({
//   __esModule: true,
//   default: jest.fn(() => ({
//     journals: [],
//     fetchJournals: jest.fn(),
//   })),
// }))

// jest.mock('../store/useAuthStore', () => ({
//   __esModule: true,
//   default: jest.fn(() => ({
//     user: { uid: 'test-uid' },
//   })),
// }))

// jest.mock('../components/JournalCard', () => 'JournalCard')

// describe('JournalScreen', () => {
//   it('component is mounted and visible to the user', () => {
//     const { getByText } = render(<JournalScreen />)
//     expect(getByText('Browse by Journals')).toBeTruthy()
//     // Assuming 'Browse by Journals' is a text visible when component mounts
//   })
// })

import React from 'react'
import { render, fireEvent } from '@testing-library/react-native'
import JournalScreen from './JournalScreen' // Adjust the import path as needed.
import { useNavigation } from '@react-navigation/native'

// Mocks setup
const mockNavigate = jest.fn()
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'), // Import and spread the actual library
  useNavigation: () => ({
    navigate: mockNavigate,
  }),
}))

jest.mock('../store/useJournalStore', () => ({
  __esModule: true,
  default: () => ({
    journals: [
      {
        id: '1',
        title: 'Test Journal',
        description: 'This is a test description',
        imageUrl: 'https://placehold.it/300x300',
        createdAt: new Date('2021-01-01'),
      },
    ],
    fetchJournals: jest.fn(),
  }),
}))

jest.mock('../store/useAuthStore', () => ({
  __esModule: true,
  default: () => ({
    user: { uid: 'test-uid' },
  }),
}))

jest.mock('../components/JournalCard', () => 'JournalCard')

describe('JournalScreen', () => {
  it('renders correctly and matches snapshot', () => {
    const tree = render(<JournalScreen />)
    expect(tree).toMatchSnapshot()
  })

  it('component is mounted and visible to the user', () => {
    const { getByText } = render(<JournalScreen />)
    expect(getByText('Browse by Journals')).toBeTruthy()
  })

  it('shows the search input', () => {
    const { getByPlaceholderText } = render(<JournalScreen />)
    expect(getByPlaceholderText('Search...')).toBeTruthy()
  })

  it('navigates to JournalEntry on "Add Journal" button press', () => {
    const { getByText } = render(<JournalScreen />)
    fireEvent.press(getByText('Add Journal'))
    expect(mockNavigate).toHaveBeenCalledWith('JournalEntry')
  })
})
