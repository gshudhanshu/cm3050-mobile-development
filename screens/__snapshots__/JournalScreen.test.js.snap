// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JournalScreen renders correctly and matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#03174C",
      "flex": 1,
      "overflow": "hidden",
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "color": "#03174C",
        "flex": 1,
        "fontFamily": "Poppins_400Regular",
        "justifyContent": "center",
        "marginHorizontal": 39,
        "marginVertical": 39,
      }
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "paddingBottom": 157,
        }
      }
    >
      <View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 20,
              "width": "100%",
            }
          }
        >
          <View
            style={
              {
                "height": 69,
                "width": 69,
              }
            }
          />
          <Text
            style={
              [
                {
                  "fontFamily": "Poppins_400Regular",
                },
                {
                  "color": "#FFFFFF",
                  "fontSize": 39,
                },
              ]
            }
          >
            Journal
          </Text>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 69,
                "opacity": 1,
                "width": 69,
              }
            }
            testID="profile-avatar"
          >
            <Image
              source={
                {
                  "uri": "https://placehold.co/100x100.png",
                }
              }
              style={
                [
                  {
                    "borderRadius": 98,
                  },
                  {
                    "height": 69,
                    "width": 69,
                  },
                ]
              }
            />
          </View>
        </View>
        <View
          style={
            {
              "marginVertical": 20,
            }
          }
        >
          <TextInput
            onChangeText={[Function]}
            placeholder="Search..."
            style={
              [
                {
                  "backgroundColor": "#EBEAEC",
                  "borderRadius": 20,
                  "fontFamily": "Poppins_400Regular",
                  "fontSize": 26.400000000000002,
                  "padding": 31,
                },
              ]
            }
            value=""
          />
        </View>
        <View>
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginVertical": 29,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "fontFamily": "Poppins_600SemiBold",
                  },
                  {
                    "color": "#FFFFFF",
                    "fontSize": 31,
                  },
                ]
              }
            >
              Browse by Journals
            </Text>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "backgroundColor": "#1DAC92",
                  "borderRadius": 100,
                  "opacity": 1,
                  "padding": 31,
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "fontFamily": "Poppins_400Regular",
                    },
                    [
                      {
                        "color": "#FFFFFF",
                        "fontSize": 24,
                        "textAlign": "center",
                        "textTransform": "uppercase",
                      },
                    ],
                  ]
                }
              >
                Add Journal
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "gap": 39,
              }
            }
          >
            <JournalCard
              date="January 01, 2021"
              description="This is a test description"
              imageUri="https://placehold.it/300x300"
              onButtonPress={[Function]}
              title="Test Journal"
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
